--- __tmp/debian/changelog.orig	2023-10-16 16:58:35.319963597 +0300
+++ __tmp/debian/changelog	2023-10-16 16:58:35.935963590 +0300
@@ -1,3 +1,17 @@
+network-manager-applet (1.20.0-3+ci202310161658+astra7) unstable; urgency=medium
+
+  *  Add a notification about NetworkManager has detected an IP address
+    conflict (BT-21199).
+  *  workaround for invisibale tray icon when Gdk/WindowScalingFactor >
+    1 (BT-25391)
+  *  fix margins in info dialog (BT-17377)
+  *  desktop entry translation (BT-15076)
+  *  fix new connection dialog black margins (BT-9836)
+  *  tray icon theme
+  *  Fix meson.build buildfix
+
+ -- Aleksey V. Kolin <support@astralinux.ru>  Mon, 16 Oct 2023 16:58:35 +0300
+
 network-manager-applet (1.20.0-3) unstable; urgency=medium
 
   * applet: support activating WireGuard connections as VPNs (Closes: #983641)
