--- a/src/connection-editor/nm-connection-editor.c	2022-10-06 13:03:45.000000000 +0300
+++ b/src/connection-editor/nm-connection-editor.c	2024-04-11 13:38:06.667973304 +0300
@@ -153,6 +153,8 @@
 		sensitive = ce_polkit_button_get_authorized (CE_POLKIT_BUTTON (editor->ok_button));
 	}
 
+	gtk_widget_set_sensitive (editor->ok_button, sensitive && !editor->last_validation_error);
+
 	/* Cancel button is always sensitive */
 	gtk_widget_set_sensitive (GTK_WIDGET (editor->cancel_button), TRUE);
 
